### DISCUSSION TITLE
How to change URL?

### DISCUSSION THREAD

**Discussion by Century1026** [inventree/InvenTree#7308] [Q&A] [answered]
I made changes here, but when I rebuild the project, server cant boot up for some errors. is this the correct way to modify url?
![image](https://github.com/inventree/InvenTree/assets/106154819/a276c570-dd43-4802-9a1e-48a8d2cae857)

**Comment by wolflu05** [comment]
Please provide more information on what you're trying to do. It seems like you're using the development environment. I would recommend to start with the [devcontainer](https://docs.inventree.org/en/stable/develop/devcontainer/), as with them it's really easy to spin up an development environment. But from your change with the site url, I'm thinking you want to run a production setup. So please follow [these docs](https://docs.inventree.org/en/stable/start/docker_install/). They guide you on how to get started. If anything is still unclear then, feel free to comment again.

**Reply by Century1026** [reply]
Hi,
Thank you for your prompt response. I'm trying to costomize the URL, so that not only local user can access the system over the Internet.

**Reply by Century1026** [reply]
I'm acturally running devcontainer:)
![image](https://github.com/inventree/InvenTree/assets/106154819/fa966a9e-655f-4a7b-8f29-1ee32854ac75)

**Comment by SchrodingersGat** [comment]
> server cant boot up for some errors.

When making a question here, please include any error messages you are seeing. Otherwise there is no way we can help you out.

**Reply by Century1026** [reply]
After encountering the errors, I restored changes in docker-compose.yml and restarted my computer, then rebuilt dev-container again and no errors occurred. I just wanna make sure if this is the only way to change site URL because the huge amount of time of building the project is so annoying :( 
If it is, I would write down the error messages and ask here again.

**Comment by Century1026** [comment]
Here's ouput when I rebuild devcontainer after making changes in docker-compose.yaml.
Honestly I don't think this can make others access system over Internet by simply altering localhost to another

Python version 3.11.8 - /home/inventree/dev/venv/bin/python3
Traceback (most recent call last):
  File "/usr/lib/python3.11/site-packages/django/core/validators.py", line 130, in __call__
    super().__call__(value)
  File "/usr/lib/python3.11/site-packages/django/core/validators.py", line 55, in __call__
    raise ValidationError(self.message, code=self.code, params={"value": value})
django.core.exceptions.ValidationErrorPython version 3.11.8 - /home/inventree/dev/venv/bin/python3
: <exception str() failed>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/inventree/src/backend/InvenTree/manage.py", line 24, in <module>
    main()
  File "/home/inventree/src/backend/InvenTree/manage.py", line 20, in main
    execute_from_command_line(sys.argv)
  File "/usr/lib/python3.11/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
    utility.execute()
  File "/usr/lib/python3.11/site-packages/django/core/management/__init__.py", line 382, in execute
    settings.INSTALLED_APPS
  File "/usr/lib/python3.11/site-packages/django/conf/__init__.py", line 102, in __getattr__
    self._setup(name)
  File "/usr/lib/python3.11/site-packages/django/conf/__init__.py", line 89, in _setup
    self._wrapped = Settings(settings_module)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/django/conf/__init__.py", line 217, in __init__
    mod = importlib.import_module(self.SETTINGS_MODULE)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/home/inventree/src/backend/InvenTree/InvenTree/settings.py", line 1026, in <module>
    validator(SITE_URL)
  File "/usr/lib/python3.11/site-packages/django/core/validators.py", line 140, in __call__
    super().__call__(url)
  File "/usr/lib/python3.11/site-packages/django/core/validators.py", line 55, in __call__
    raise ValidationError(self.message, code=self.code, params={"value": value})
django.core.exceptions.ValidationErrorPython version 3.11.8 - /home/inventree/dev/venv/bin/python3
: <exception str() failed>

**Reply by SchrodingersGat** [reply]
@Century1026 two questions:

### SITE_URL

> File "/home/inventree/src/backend/InvenTree/InvenTree/settings.py", line 1026, in
validator(SITE_URL)

This would indicate that the `SITE_URL` is invalid (for some reason). Can you share what you are trying to set it to?

### Devcontainer

> Honestly I don't think this can make others access system over Internet by simply altering localhost to another

Are you trying to get others to access a system you are running in the *devcontainer* setup? That is *only* intended for devlopment and not production, I doubt it would work in that way at all...

**Reply by Century1026** [reply]
1. I made it my username: "century1026" lol
![image](https://github.com/inventree/InvenTree/assets/106154819/c2547aac-bec1-41c8-afd6-11579ce0a55e)

2. Altight, I'll try to figure out how production work, thanks!

**Reply by SchrodingersGat** [reply]
>  made it my username: "century1026"

That's.. not going to work. You have to point it to a location that is resovable by DNS

**Reply by Century1026** [reply]
Yeah, you're right, URL like that requires DNS.
I configured production successfully but it's like running on original static code. Is it possible to turn modified code under development into production so that people over Internet can access my redeveloped system?

**Reply by SchrodingersGat** [reply]
@Century1026 yes, but requires some tweaking to get working. 

If you have a DNS pointing to the server where you're running the devcontainer docker setup, it should work. However it is up to you to ensure that the routing is working correctly (webserver setup is outside the scope of what we'd normally help debug on here)...

**Comment by Century1026** [comment]
@wolflu05 @SchrodingersGat 
I used ngrok to create a url with port 80 forwarded and added port 80 on firewall. I also changed devcontainer.json and docker-compose.yml and rebuilt dev-container. It generated a link 0.0.0.0:80, but I failed to access it even on my computer.
I know it's out of the scope of questioning. Any suggestions would help, please.
![image](https://github.com/inventree/InvenTree/assets/106154819/99e59381-5373-4a54-8daf-fe264cfb807a)
![image](https://github.com/inventree/InvenTree/assets/106154819/6dcfe56f-efdc-430f-a473-9933d223e014)
![image](https://github.com/inventree/InvenTree/assets/106154819/06fad6b8-102a-49d3-bae6-aeac37873fc7)

**Reply by SchrodingersGat** [reply, marked as answer]
Is it possible that you have another server already running on port 80?

I tried the following, from the "stock" devcontainer config (i.e. no changes out of the box).

From the command line, run `invoke server -a 0.0.0.0:8888`

Then, the server is available on port `8888` (change as per your requirements). No other adjustments should be needded.

**Reply by Century1026** [reply]
> Is it possible that you have another server already running on port 80?
> 
> I tried the following, from the "stock" devcontainer config (i.e. no changes out of the box).
> 
> From the command line, run `invoke server -a 0.0.0.0:8888`
> 
> Then, the server is available on port `8888` (change as per your requirements). No other adjustments should be needded.

Thank you so much! It worked!!
However, I encountered the same login issue as before while logging in from the new forwarded url.
![image](https://github.com/inventree/InvenTree/assets/106154819/678faf16-6902-4217-ab2a-299efd783cf7)
This is the last step, I'm so excited!

**Reply by Century1026** [reply]
Response:
```




<!DOCTYPE html>
<html lang="en">
<head>

<!-- Required meta tags -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">

<!-- Favicon -->
<link rel="apple-touch-icon" sizes="57x57" href="/static/img/favicon/apple-icon-57x57.png">
<link rel="apple-touch-icon" sizes="60x60" href="/static/img/favicon/apple-icon-60x60.png">
<link rel="apple-touch-icon" sizes="72x72" href="/static/img/favicon/apple-icon-72x72.png">
<link rel="apple-touch-icon" sizes="76x76" href="/static/img/favicon/apple-icon-76x76.png">
<link rel="apple-touch-icon" sizes="114x114" href="/static/img/favicon/apple-icon-114x114.png">
<link rel="apple-touch-icon" sizes="120x120" href="/static/img/favicon/apple-icon-120x120.png">
<link rel="apple-touch-icon" sizes="144x144" href="/static/img/favicon/apple-icon-144x144.png">
<link rel="apple-touch-icon" sizes="152x152" href="/static/img/favicon/apple-icon-152x152.png">
<link rel="apple-touch-icon" sizes="180x180" href="/static/img/favicon/apple-icon-180x180.png">
<link rel="icon" type="image/png" sizes="192x192"  href="/static/img/favicon/android-icon-192x192.png">
<link rel="icon" type="image/png" sizes="32x32" href="/static/img/favicon/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="96x96" href="/static/img/favicon/favicon-96x96.png">
<link rel="icon" type="image/png" sizes="16x16" href="/static/img/favicon/favicon-16x16.png">
<link rel="manifest" href="/static/img/favicon/manifest.json">
<meta name="msapplication-TileColor" content="#ffffff">
<meta name="msapplication-TileImage" content="/static/img/favicon/ms-icon-144x144.png">
<meta name="theme-color" content="#ffffff">


<!-- CSS -->
<link rel="stylesheet" href="/static/fontawesome/css/brands.css">
<link rel="stylesheet" href="/static/fontawesome/css/solid.css">
<link rel="stylesheet" href="/static/bootstrap/css/bootstrap.min.css">
<link rel="stylesheet" href="/static/select2/css/select2.css">
<link rel="stylesheet" href="/static/select2/css/select2-bootstrap-5-theme.css">
<link rel="stylesheet" href="/static/css/inventree.css">

<link rel="stylesheet" href="/static/css/color-themes/default.css">

<title>
    Inventree | 
</title>



</head>

<body class='login-screen' style='background: url(/static/img/inventree_splash.jpg); background-size: cover;'>

    <div class='container-fluid'>
        <div class='notification-area' id='alerts'>
            <!-- Div for displayed alerts -->
        </div>
    </div>

    <div class='main body-wrapper login-screen d-flex'>

        <div class='login-container'>
        <div class="row">
            <div class='container-fluid'>

                <div class='clearfix content-heading login-header d-flex flex-wrap'>
                    <img class="pull-left" src="/static/img/inventree.png" alt='InvenTree logo' width="60" height="60"/>
                    <span class='flex' style='flex-grow: 1;'></span>

                    <span class='float-right'><h3>Inventree</h3></span>
                </div>
            </div>
                <div class='container-fluid'>
                    <hr>
                    
<h3>Authentication Failure</h3>

<div class='alert alert-danger alert-block'>
    You have been logged out from InvenTree.
</div>
<hr>
<div class='btn-group float-right' role='group'>
    <a type='button' class='btn btn-primary' href='/accounts/login/'>
        <span class='fas fa-sign-in-alt'></span> Login
    </a>
</div>


                </div>
        </div>

        </div>

        
        
    </div>



<!-- general JS -->


<!-- jquery -->
<script type="text/javascript" src="/static/script/jquery_3.3.1_jquery.min.js"></script>
<script type='text/javascript' src="/static/script/jquery.form.min.js"></script>
<script type='text/javascript' src="/static/script/jquery-ui/jquery-ui.min.js"></script>

<!-- Bootstrap-->
<script type="text/javascript" src="/static/bootstrap/js/bootstrap.bundle.min.js"></script>

<!-- Bootstrap Table -->
<script defer type='text/javascript' src="/static/script/bootstrap/bootstrap-treeview.js"></script>
<script defer type='text/javascript' src='/static/treegrid/js/jquery.treegrid.js'></script>
<script defer type='text/javascript' src='/static/treegrid/js/jquery.treegrid.bootstrap3.js'></script>
<script defer type='text/javascript' src="/static/bootstrap-table/bootstrap-table.min.js"></script>
<script defer type='text/javascript' src='/static/bootstrap-table/extensions/group-by-v2/bootstrap-table-group-by.min.js'></script>
<script defer type='text/javascript' src='/static/bootstrap-table/extensions/filter-control/bootstrap-table-filter-control.min.js'></script>
<script defer type='text/javascript' src='/static/bootstrap-table/extensions/treegrid/bootstrap-table-treegrid.min.js'></script>
<script defer type='text/javascript' src='/static/bootstrap-table/extensions/custom-view/bootstrap-table-custom-view.min.js'></script>

<!-- fontawesome -->
<script defer type='text/javascript' src="/static/fontawesome/js/solid.min.js"></script>
<script defer type='text/javascript' src="/static/fontawesome/js/regular.min.js"></script>
<script defer type='text/javascript' src="/static/fontawesome/js/brands.min.js"></script>
<script defer type='text/javascript' src="/static/fontawesome/js/fontawesome.min.js"></script>

<!-- 3rd party general js -->
<script defer type="text/javascript" src="/static/fullcalendar/main.min.js"></script>
<script defer type="text/javascript" src="/static/fullcalendar/locales-all.min.js"></script>
<script defer type="text/javascript" src="/static/select2/js/select2.full.min.js"></script>
<script defer type='text/javascript' src="/static/script/moment.js"></script>
<script defer type='text/javascript' src="/static/script/chart.js"></script>
<script defer type='text/javascript' src="/static/script/chartjs-adapter-moment.js"></script>
<script defer type='text/javascript' src="/static/script/clipboard.min.js"></script>
<script defer type='text/javascript' src="/static/easymde/easymde.min.js"></script>
<script defer type='text/javascript' src="/static/script/randomColor.min.js"></script>
<script defer type='text/javascript' src="/static/script/html5-qrcode.min.js"></script>
<script defer type='text/javascript' src="/static/script/qrcode.min.js"></script>


<script type='text/javascript' src='/static/script/inventree/inventree.js'></script>
<script type='text/javascript' src='/static/script/inventree/message.js'></script>

<script type='text/javascript'>

$(document).ready(function () {

    

    showCachedAlerts();

    // Add brand icons for SSO providers, if available
    $('.socialaccount_provider').each(function(i, obj) {
        var el = $(this);
        var tag = el.attr('brand_name');

        var icon = window.FontAwesome.icon({prefix: 'fab', iconName: tag});

        if (icon) {
            el.prepend(`<span class='fab fa-${tag}'></span>&nbsp;`);
        }
    });

});

</script>

</body>
</html>

```

**Reply by SchrodingersGat** [reply]
What IP are you accessing the site from? You will need to add the IP to the "ALLOWED_HOSTS" and "CSRF_TRUSTED_ORIGINS" config values:

https://docs.inventree.org/en/stable/start/config/#server-access

**Reply by Century1026** [reply]
Thank you so much! I finally made it!
Oliver you are goat!
